From a4ff0e3c6dc9e8a7d02c6bbb8a818897738c6e1d Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <shengjiu.wang@freescale.com>
Date: Wed, 3 Feb 2016 09:52:26 +0800
Subject: [PATCH 1546/1691] MLK-12366: ASoC: dmaengine: fix sai underrun issue
 after suspend/resume

commit a4ff0e3c6dc9e8a7d02c6bbb8a818897738c6e1d from
git://git.freescale.com/imx/linux-2.6-imx.git imx_4.1.15_1.0.0_ga

We use the devm_snd_dmaengine_pcm_register() for platform,
don't use devm_snd_soc_register_platform(), their close
funtion is different. One will release the dma channel,
but another one won't release channel.

When system enter suspend mode, alsa library is in draining
state, it will not resume alsa driver in system resume, just
call close directly. But the alsa driver is in state
SNDRV_PCM_STATE_SUSPENDED, which is not a running state, so
the trigger stop is not called, and dmaengine_terminate_all()
is not called, then dma resource is not released. Which cause
that in the next playback the sdma driver will not be started.
Then there is a lot of SAI underrun error.

If we use the devm_snd_soc_register_platform() for platfrom
driver, there will be no issue, because it will release dma
channel, release dma resource. But we use the
devm_snd_dmaengine_pcm_register() for platform driver. So this
patch is to add dmaengine_terminate_all() in the
snd_dmaengine_pcm_close() to fix this issue.

Signed-off-by: Shengjiu Wang <shengjiu.wang@freescale.com>
(cherry picked from commit ff62968ea8c8b5414f3f049e61addcd4d2bec3b9)
---
 sound/core/pcm_dmaengine.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/sound/core/pcm_dmaengine.c b/sound/core/pcm_dmaengine.c
index 29ce8ef..e247583 100644
--- a/sound/core/pcm_dmaengine.c
+++ b/sound/core/pcm_dmaengine.c
@@ -5,7 +5,7 @@
  *  Based on:
  *	imx-pcm-dma-mx2.c, Copyright 2009 Sascha Hauer <s.hauer@pengutronix.de>
  *	mxs-pcm.c, Copyright (C) 2011 Freescale Semiconductor, Inc.
- *	imx-pcm-dma.c, Copyright (C) 2014-2015 Freescale Semiconductor, Inc.
+ *	imx-pcm-dma.c, Copyright (C) 2014-2016 Freescale Semiconductor, Inc.
  *	ep93xx-pcm.c, Copyright (C) 2006 Lennert Buytenhek <buytenh@wantstofly.org>
  *		      Copyright (C) 2006 Applied Data Systems
  *
@@ -344,6 +344,7 @@ int snd_dmaengine_pcm_close(struct snd_pcm_substream *substream)
 	struct dmaengine_pcm_runtime_data *prtd = substream_to_prtd(substream);
 
 	dma_sync_wait_tasklet(prtd->dma_chan);
+	dmaengine_terminate_all(prtd->dma_chan);
 
 	dmaengine_synchronize(prtd->dma_chan);
 	kfree(prtd);
-- 
1.9.1

